### ISSUE TITLE
aws_autoscaling_group fails when increasing number of instances

### ISSUE THREAD

**Issue by bnm22** [hashicorp/terraform-provider-aws#5241]
I recently tried to bump the number of instances in an AWS autoscaling group resource and ran into this bug. 

**Version:**
Terraform v0.11.7
provider.aws v1.27.0

`Error: Error applying plan:

1 error(s) occurred:

* module.app.aws_autoscaling_group.asg: 1 error(s) occurred:

* aws_autoscaling_group.asg: Error waiting for AutoScaling Group Capacity: "82766e2f-new-app20180718145707823100000005": Waiting up to 30m0s: Need exactly 1 healthy instances in ASG, have 3. Most recent activity: {
  ActivityId: "134592b7-c695-f45e-8cef-b3ef2fc30b7c",
  AutoScalingGroupName: "82766e2f-new-app20180718145707823100000005",
  Cause: "At 2018-07-18T15:10:27Z a user request update of AutoScalingGroup constraints to min: 3, max: 3, desired: 3 changing the desired capacity from 1 to 3.  At 2018-07-18T15:10:50Z an instance was started in response to a difference between desired and actual capacity, increasing the capacity from 1 to 3.",
  Description: "Launching a new EC2 instance: i-0cd19a8b939a11eeb",
  Details: "{\"Subnet ID\":\"subnet-1afesfsfseds\",\"Availability Zone\":\"us-east-1a\"}",
  EndTime: 2018-07-18 15:11:25 +0000 UTC,
  Progress: 100,
  StartTime: 2018-07-18 15:10:52.285 +0000 UTC,
  StatusCode: "Successful"
}`

**Comment by smastrorocco** [comment]
Seeing this issue on Terraform v0.11.7 provider 1.42.0

```
module.app.aws_autoscaling_group.asg: Modifying... (ID: xxxxxxxx) 
November 6th 2018 18:06:59Info
  max_size:              "3" => "6" 
November 6th 2018 18:06:59Info
  min_size:              "3" => "6" 
November 6th 2018 18:06:59Info
  wait_for_elb_capacity: "3" => "6" 
....
Error: Error applying plan: 
November 6th 2018 18:37:01Error
1 error(s) occurred: 
November 6th 2018 18:37:01Error
* module.app.aws_autoscaling_group.asg: 1 error(s) occurred: 
November 6th 2018 18:37:01Error
* aws_autoscaling_group.asg: Error waiting for AutoScaling Group Capacity: "xxxxxxxx": Waiting up to 30m0s: Need exactly 3 healthy instances in ASG, have 6
```

**Comment by smastrorocco** [comment]
We are still seeing this issue:

```
...
max_size:              "8" => "10" 
min_size:              "8" => "10"
...
Need exactly 8 healthy instances in ASG, have 10.
```

**Comment by lachlancooper** [comment]
I saw this issue today with Terraform v0.12.18, AWS provider v2.48.0, while trying to increase `min_size` and `wait_for_elb_capacity` from 2 to 3. The relevant section of apply output (after 2 hours of `Still modifying...`), is:

```
Error: Error waiting for AutoScaling Group Capacity: "cluster-foo": Waiting up to 2h0m0s: Need exactly 2 healthy instances in ASG, have 3. Most recent activity: {
...
Cause: "At 2020-02-10T00:48:07Z a user request update of AutoScalingGroup constraints to min: 3, max: 4, desired: 3 changing the desired capacity from 2 to 3.  At 2020-02-10T00:48:14Z an instance was started in response to a difference between desired and actual capacity, increasing the capacity from 2 to 3.",
...
```

The `Need exactly <x> healthy instances in ASG` message is referencing the existing instance count, not the new one. That specific message comes from the first `if` block [in the capacitySatisfiedUpdate() function](https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_autoscaling_group_waiting.go#L189). The problem is with `wantASG`, which comes from the `desired_capacity` schema key.

The `desired_capacity` resource argument is optional and will default to `min_size` if not specified. I didn't set or change it on the Terraform resource, but because it's a property of the AWS resource it has a defined value. Terraform doesn't seem to be aware it will change, since the plan output shows the existing value for `desired_capacity`:

```
Terraform will perform the following actions:

  # aws_autoscaling_group.cluster-foo will be updated in-place
  ~ resource "aws_autoscaling_group" "cluster-foo" {
...
        default_cooldown          = 300
        desired_capacity          = 2
        enabled_metrics           = []
...
        max_instance_lifetime     = 0
      ~ max_size                  = 3 -> 4
        metrics_granularity       = "1Minute"
      ~ min_size                  = 2 -> 3
...
        wait_for_capacity_timeout = "2h"
      ~ wait_for_elb_capacity     = 2 -> 3

        tag {
...
```

The [field is marked as Computed](https://github.com/terraform-providers/terraform-provider-aws/blob/master/aws/resource_aws_autoscaling_group.go#L205), but that apparently only applies on creation, not update.

I suspect a fix could be to check `d.HasChange("desired_capacity")` and, if not, skip the first `if` block in `capacitySatisfiedUpdate()`. I don't understand Terraform internals well though so perhaps there's a more fundamental issue with desired/min/max capacity calculation.

**Comment by KMahankali** [comment]
Is this issue fixed? I am still seeing the below error when tried to increase the min and max number

![image](https://user-images.githubusercontent.com/24401562/91414568-c0d1f380-e81a-11ea-8b05-3d26c4413144.png)

**Comment by lachlancooper** [comment]
> Is this issue fixed? I am still seeing the below error when tried to increase the min and max number

No. I've proposed a fix in #12018 but it hasn't been reviewed yet.

**Comment by rtizzy** [comment]
Can someone from the Terraform team take a few moments to triage this issue and the associated pull request? Looks like this has been open for a long time. 

For what it's worth, I'm running into this issue when decreasing from 1 > 0 and while it works for one autoscaling group with no issues, another virtually identical autoscaling group ends up with the behavior mentioned previously (Waiting till default timeout then failing). 

```
│ Error: error waiting for Auto Scaling Group Capacity: "REDACTED-dev-blue-api-asg20210923145426274700000003": Waiting up to 10m0s: Need exactly 1 healthy instances in ASG, have 0. Most recent activity: {
│   ActivityId: "c5b173ea-bc9b-41d9-835a-d0ba59526925",
│   AutoScalingGroupARN: "arn:aws:autoscaling:us-west-2:897744604563:autoScalingGroup:cc808904-cda2-42ee-a535-c6a9fb6ec9f2:autoScalingGroupName/REDACTED-dev-blue-api-asg20210923145426274700000003",
│   AutoScalingGroupName: "REDACTED-dev-blue-api-asg20210923145426274700000003",
│   Cause: "At 2021-09-23T15:27:15Z a user request update of AutoScalingGroup constraints to min: 0, max: 0, desired: 0 changing the desired capacity from 1 to 0.  At 2021-09-23T15:27:19Z an instance was taken out of service in response to a difference between desired and actual capacity, shrinking the capacity from 1 to 0.  At 2021-09-23T15:27:19Z instance i-01d4443697fea5b6d was selected for termination.",
│   Description: "Terminating EC2 instance: i-01d4443697fea5b6d",
│   Details: "{\"Subnet ID\":\"subnet-REDACTED\",\"Availability Zone\":\"us-west-2c\"}",
│   EndTime: 2021-09-23 15:34:21 +0000 UTC,
│   Progress: 100,
│   StartTime: 2021-09-23 15:27:19.437 +0000 UTC,
│   StatusCode: "Successful"
│ }
│ 
│   with aws_autoscaling_group.api_blue,
│   on asg_api.tf line 117, in resource "aws_autoscaling_group" "api_blue":
│  117: resource "aws_autoscaling_group" "api_blue" {
│ 
╵
Releasing state lock. This may take a few moments...
ERRO[0648] 1 error occurred:
        * exit status 1
```
